components/engine_traits/src/properties.rs

```diff
@@ -0,0 +1,83 @@
+// Copyright 2017 TiKV Project Authors. Licensed under Apache-2.0.
+
+use std::collections::BTreeMap;
+use std::io::Read;
+use std::ops::{Deref, DerefMut};
+use tikv_util::codec::number::{self, NumberEncoder};
+use tikv_util::codec::Result;
+
+#[derive(Clone, Debug, Default)]
+pub struct IndexHandle {
+    pub size: u64,   // The size of the stored block
+    pub offset: u64, // The offset of the block in the file
+}
+
+#[derive(Debug, Default)]
+pub struct IndexHandles(BTreeMap<Vec<u8>, IndexHandle>);
+
+impl Deref for IndexHandles {
+    type Target = BTreeMap<Vec<u8>, IndexHandle>;
+    fn deref(&self) -> &BTreeMap<Vec<u8>, IndexHandle> {
+        &self.0
+    }
+}
+
+impl DerefMut for IndexHandles {
+    fn deref_mut(&mut self) -> &mut BTreeMap<Vec<u8>, IndexHandle> {
+        &mut self.0
+    }
+}
+
+impl IndexHandles {
+    pub fn new() -> IndexHandles {
+        IndexHandles(BTreeMap::new())
+    }
+
+    pub fn into_map(self) -> BTreeMap<Vec<u8>, IndexHandle> {
+        self.0
+    }
+
+    pub fn add(&mut self, key: Vec<u8>, index_handle: IndexHandle) {
+        self.0.insert(key, index_handle);
+    }
+
+    // Format: | klen | k | v.size | v.offset |
+    pub fn encode(&self) -> Vec<u8> {
+        let mut buf = Vec::with_capacity(1024);
+        for (k, v) in &self.0 {
+            buf.encode_u64(k.len() as u64).unwrap();
+            buf.extend(k);
+            buf.encode_u64(v.size).unwrap();
+            buf.encode_u64(v.offset).unwrap();
+        }
+        buf
+    }
+
+    pub fn decode(mut buf: &[u8]) -> Result<IndexHandles> {
+        let mut res = BTreeMap::new();
+        while !buf.is_empty() {
+            let klen = number::decode_u64(&mut buf)?;
+            let mut k = vec![0; klen as usize];
+            buf.read_exact(&mut k)?;
```

Comment:
These two lines can be 60× faster if you manipulate the bytes directly instead of going through `Read`. (I'm not sure if `IndexHandles::decode()` is called in a hot loop)

```suggestion
            let klen = klen as usize;
            if buf.len() < klen {
                Err(std::io::ErrorKind::UnexpectedEof)?;
            }
            let (k, buf_tail) = buf.split_at(klen);
            buf = buf_tail;
            let k = k.to_vec();
```

Reply:
Thanks for the suggestion. Can we do that in a different PR though? Optimization is not the purpose of this one, and this code is just being moved around here.

Reply:
Thanks for the suggestion. Can we do that in a different PR though? Optimization is not the purpose of this one, and this code is just being moved around here.

Reply:
OK.